AUTH: move account locking functionality into mixin class

nursix opened the pull request:
- simplify session locking (session not reusable)
- cleanup and simplify overall logic

nursix committed:
AUTH: move account locking functionality into mixin class, simplify
session locking (session not reusable), simplify overall logic, cleanup
and simplify overall logic

nursix committed:
Modify email verification process to prevent key guessing

nursix commented:
With the last commit, key guessing in the verification process becomes futile:

The revised flow sends the user a link (with a key) and an activation code, which are used to produce a hash to match the `registration_key` previously stored in the account. Neither the key or the activation code are stored in the database.

That means:
a) the user will need both, key and activation code, to unlock the account - and therefore the entire verification email. The link with a key alone is no longer enough.
b) account verification is a manual process, it is not enough to open a URL - the user must also enter the activation code into an interactive form.
c) no possible hash will ever match any value of `registration_key` that locks the account, so regardless how good a guess it would be, it cannot overcome a "disabled" or "blocked".

This should resolve #14

nursix committed:
Replace locked-flag by "failed" registration key, allow re-enabling of
failed accounts from Admin-UI

awjans reviewed (commented):
Your Python-Foo is strong.

nursix committed:
AUTH: logging/notification of account unlocking made explicit

nursix commented:
Added a commit to make logging/notification of account unlocking explicit. 
- also changed to avoid the DB write during unlocking unless the record is actually changing

awjans reviewed (approved):
This PR is well architected, and seems to be the minimum amount of code to implement the functionality. The blueprint can be marked as completed and this can be merged.

nursix commented:
Thanks - merged.

I think there is one more piece of functionality in the blueprint that we could consider, which is to allow for unlocking of an account by way of email re-verification - pretty much the same way as I did it in `verify_email`, just a different end-point (or maybe even the same endpoint).

I can raise a separate issue for that, though.

awjans commented:
Would this be self-service account unlocking? Or an option for the admin to send a re-verification/unlock email instead of just unlocking the account? Would this also force a password reset?
